src/app/data-model-provider/tests/TestMetadataList.cpp

```diff
@@ -0,0 +1,396 @@
+        EXPECT_EQ(list1.Append(4), CHIP_ERROR_INCORRECT_STATE);
+
+        MetadataList<int> list2{ MetadataList<int>::FromConstSpan(Span<const int>{ kConstantArray }) };
+        EXPECT_EQ(list2.size(), 3u);
+        EXPECT_EQ(list2[0], 1);
+        EXPECT_EQ(list2[1], 2);
+        EXPECT_EQ(list2[2], 3);
+
+        auto list1Span = list1.GetSpanValidForLifetime();
+        auto list2Span = list2.GetSpanValidForLifetime();
+
+        EXPECT_EQ(list1Span.data(), list2Span.data());
+        EXPECT_EQ(list1Span.size(), list2Span.size());
+        EXPECT_EQ(list1Span.data(), kConstantArray.data());
+    }
+}
+
+enum MinCommandPrivilege : uint8_t
+{
+    kOperate = 0u,
+    kManage  = 1u,
+    kAdmin   = 2u,
+
+    kMax = kAdmin
+};
+
+static constexpr uint16_t kPrivilegeFieldWidth = 0x3;
+static constexpr uint16_t kPrivilegeFieldMask  = static_cast<uint16_t>((1u << kPrivilegeFieldWidth) - 1);
+static_assert(MinCommandPrivilege::kMax <= kPrivilegeFieldMask, "Privilege mask is not wide enough");
+
+// Bitmask values for different Command qualities.
+enum class CommandMetadataFlags : uint16_t
+{
+    kFabricScoped = 0x1 << 0,
+    kTimed        = 0x1 << 1, // `T` quality on commands
+    kLargeMessage = 0x1 << 2, // `L` quality on commands
+    kIsResponse   = 0x1 << 3, // Command is server => client response
+
+    kMinPrivilegeValueMask = kPrivilegeFieldMask << 4
+
+};
+
+typedef BitMask<CommandMetadataFlags> CommandMetadata;
+
+// MinCommandPrivilege GetMinCommandPrivilege(CommandMetadata metadata) const {
+//   return static_cast<MinCommandPrivilege>(metadata.GetField(CommandMetadataFlags::kMinPrivilegeValueMask));
+// }
+
+struct CommandEntry
+{
+    CommandId mei;
+    CommandMetadata metadata;
+};
+
+enum class NetworkCommissioningFeatureBits : uint8_t
+{
+    // TODO: NOT REAL VALUEs
+    kWifi     = 1 << 0,
+    kThread   = 1 << 1,
+    kEthernet = 1 << 2,
+};
+
+enum NetworkCommissioningCommands : CommandId
+{
+    kScanNetworks             = 0x00u, // | client => server | ScanNetworksResponse   | A      | WI \| TH
+    kScanNetworksResponse     = 0x01u, // | client <= server | N                      |        | WI \| TH
+    kAddOrUpdateWiFiNetwork   = 0x02u, // | client => server | NetworkConfigResponse  | A      | WI
+    kAddOrUpdateThreadNetwork = 0x03u, // | client => server | NetworkConfigResponse  | A      | TH
+    kRemoveNetwork            = 0x04u, // | client => server | NetworkConfigResponse  | A      | WI \| TH
+    kNetworkConfigResponse    = 0x05u, // | client <= server | N                      |        | WI \| TH
+    kConnectNetwork           = 0x06u, // | client => server | ConnectNetworkResponse | A      | WI \| TH
+    kConnectNetworkResponse   = 0x07u, // | client <= server | N                      |        | WI \| TH
+    kReorderNetwork           = 0x08u, // | client => server | NetworkConfigResponse  | A      | WI \| TH
+
+};
+
+static const std::array<CommandEntry, 9> kAllCommands{
```

Comment:
Can you check if this is run rather a flat constexpr table? I.e. is `SetField` seen called in an initializer?

Reply:
Why?